### PR TITLE
feat: implement boundInt

### PR DESCRIPTION
**Pull request by JhChoy**
Implementing `boundInt(int256,int256,int256)` at StdUtils.sol to help fuzzing test with integer arguments.

One problem is that `log(string,int256)` does not exist so I used `log(string,uint256)` instead of it.

### CODE REVIEW

File: test/StdUtils.t.sol
Comment (mds1) [inline test/StdUtils.t.sol:101]: Can we also tests for the inverse cases, e.g.
```solidity
assertEq(boundInt(type(int256).min, 10, 150), 10);
assertEq(boundInt(type(int256).min + 1, 10, 150), 11);
// add the other two
assertEq(boundInt(type(int256).max, -50, -10), -10);
assertEq(boundInt(type(int256).max - 1, -50, -10), -11);
// add the other two
```

Maybe good to explicitly test when zero is both the lower and upper bound in those scenarios also. Ints have some more edge cases than Uints IME so just want to make sure it's well tested 🙂

File: src/StdUtils.sol
Comment (mds1) [inline src/StdUtils.sol:7]: _(arbitrary comment location to start a thread)_ 

> One problem is that `log(string,int256)` does not exist so I used `log(string,uint256)` instead of it.

If you want to add support for this:
1. Add the required method to `console2.sol` in forge-std
2. Update the [ABI file](https://github.com/foundry-rs/foundry/blob/master/evm/abi/console.json) in foundry

I think that should be all that's needed, but I may be missing something. Also related is https://github.com/foundry-rs/forge-std/issues/129 if you wanted to add that overload at the same time

File: src/StdUtils.sol
Comment (mds1) [inline src/StdUtils.sol:49]: Mind just adding a comment explaining how this works / linking to a reference? Want to make sure it's clear/understandable to maintainers (including myself) 😅

I think what you're doing is shifting all values from the int domain to the uint domain, running `_bound`, then shifting them back? Which makes sense and is how I'd do it, just not intuitive how the shift works

File: src/StdUtils.sol
Comment (mds1) [inline src/StdUtils.sol:44]: Let's just call this `bound`
```suggestion
    function bound(int256 x, int256 min, int256 max) internal view virtual returns (int256 result) {
```

File: src/StdUtils.sol
Comment (mds1) [inline src/StdUtils.sol:11]: One side effect here: by removing this `require` statement, it's no longer safe to use `_bound` as an alternative to `bound` to get the same behavior without the log noise, so this is technically a breaking change that we should avoid IMO. I think we can keep this here, and just leave the new require statement in the `boundInt` method. You get a duplicate check as a result, but that's ok

File: src/StdUtils.sol
Comment (mds1) [inline src/StdUtils.sol:52]: This shift back is more intuitive, but maybe worth a comment here anyway

**Review by mds1 — COMMENTED** [review]
This is awesome, thanks a lot! Pinging @paulrberg since I know he was interesting in this method.

Just a few nits for you 🙂

File: test/StdUtils.t.sol
Comment (JhChoy) [reply test/StdUtils.t.sol:101]: Great! I'll add them.

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:7]: Okay, I'll check this out.

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:49]: Agree, I'll add some comments about this code.

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:44]: I wanted to... But Solidity says it is duplicated with `bound(uint256,uint256,uint256)` 😭.

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:11]: That's a good point. I'll fix it.

File: src/StdUtils.sol
Comment (mds1) [reply src/StdUtils.sol:44]: We can call it `bound`, because it has a different function signature than the other `bound`, so that's an allowed overload.

But if you change the name to `bound` with no other changes, solc will error because certain statements like `assertEq(bound(5, 0, 4), 0), 0)` are now ambiguous—are those inputs ints or uints? It can't tell which `bound` implementation to use.

So we can resolve that by just casting the first input value so it knows which overload to use: `assertEq(bound(uint256(5), 0, 4), 0)`

This is a bit clunky for tests where we hardcode constants and need to cast everywhere now, but I think that's ok. In standard fuzz tests, solidity will know the type of that first input to bound because it a function input of uint256 or int256 in the fuzz test, so the casting won't be needed and UX will be better for users

File: src/StdUtils.sol
Comment (mds1) [inline src/StdUtils.sol:8]: _(arbitrary comment location to start a thread)_

Looks like there was also a test failure in CI:

```
[FAIL. Reason: Arithmetic over/underflow Counterexample: calldata=0x87b338ab000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000000, args=[0, -57896044618658097711785492504343953926634992332820282019728792003956564819967, -57896044618658097711785492504343953926634992332820282019728792003956564819968]] testBoundInt(int256,int256,int256) (runs: 630, μ: 7029, ~: 7054)
Traces:
  [2013] StdUtilsTest::testBoundInt(0, -57896044618658097711785492504343953926634992332820282019728792003956564819967, -57896044618658097711785492504343953926634992332820282019728792003956564819968) 
    └─ ← "Arithmetic over/underflow"
```

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol]: [328f4df](https://github.com/foundry-rs/forge-std/pull/253/commits/328f4dfe531b66876ccafa8be1184b923b04c02d)

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:56]: [328f4df](https://github.com/foundry-rs/forge-std/pull/253/commits/328f4dfe531b66876ccafa8be1184b923b04c02d)

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:8]: [c43e95b](https://github.com/foundry-rs/forge-std/pull/253/commits/c43e95be0f0771d0d188cd1b855ad0fe972b5b5c)

File: test/StdUtils.t.sol
Comment (JhChoy) [reply test/StdUtils.t.sol]: [c43e95b](https://github.com/foundry-rs/forge-std/pull/253/commits/c43e95be0f0771d0d188cd1b855ad0fe972b5b5c)

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:11]: [174b694](https://github.com/foundry-rs/forge-std/pull/253/commits/174b694c8b60d2670d5e56f6d70ca554c80cc37b)

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol]: Oh, I understand. Now it works well.
Applied at here. [25ff698](https://github.com/foundry-rs/forge-std/pull/253/commits/25ff69883e47be0c10b0e10fdbf7e391efc97acd)

File: test/StdUtils.t.sol
Comment (mds1) [reply test/StdUtils.t.sol]: Small nit, do you mind reordering the tests like this? It makes it easier to read since sequential/related tests are next to each other. (I'm happy to push the commit, but don't want to push yet in case you're working on https://github.com/foundry-rs/forge-std/pull/253#discussion_r1045104818).

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:7]: What will be the better one, `log(string,int256)` or `logInt(string,int256)`?
At the [related thread](https://github.com/NomicFoundation/hardhat/issues/1440) of #129, they decided to name `logInt(int256...)` rather than `log(int256...)` because of the ambiguity.

File: test/StdUtils.t.sol
Comment (JhChoy) [reply test/StdUtils.t.sol]: Got it, I'll apply it.

File: src/StdUtils.sol
Comment (mds1) [reply src/StdUtils.sol:7]: I'm ok with `log(string,int256)` because in practice you'll be doing `console.log(myVar)` (unambiguous since `myVar` has a type) instead of `console.log(3)` (ambiguous because we don't know 3's type)

File: test/StdUtils.t.sol
Comment (JhChoy) [reply test/StdUtils.t.sol]: Fixed at [here](https://github.com/foundry-rs/forge-std/pull/253/commits/3e50c5d244cdfa055aa6468334ad6193e746d7fe).

File: src/StdUtils.sol
Comment (JhChoy) [reply src/StdUtils.sol:7]: Applied at [here](https://github.com/foundry-rs/forge-std/pull/253/commits/3e50c5d244cdfa055aa6468334ad6193e746d7fe)!

**Comment by paulrberg** [comment]
I don't currently have bandwidth to review but this is good stuff, thanks for implementing a signed bound, @JhChoy! I needed this for my [PRBMath](prb-math) tests.

File: src/StdUtils.sol
Comment (mds1) [reply src/StdUtils.sol:7]: Thanks! Note that the logs won't actually be shown until we upstream the corresponding change to forge. I can do that soon though

**Review by mds1 — APPROVED** [review]
LGTM, thanks @JhChoy! 

@ZeroEkkusu mind taking a look before we merge?

**Comment by mds1** [comment]
@JhChoy just a heads up I pushed two small commits to this branch and opened https://github.com/foundry-rs/foundry/issues/3884

**Comment by JhChoy** [comment]
@mds1 Thanks for the review! 🙌

**Review by ZeroEkkusu — APPROVED** [review]
Thanks.

**Comment by ZeroEkkusu** [comment]
Pushed changes for `1.2.0` release.

After https://github.com/foundry-rs/foundry/issues/3884 is resolved, we will remove the workaround.

Note: `console2_log` functions are part of the upcoming "remove Components" PR, so I made those changes here to avoid conflicts.

File: src/StdUtils.sol
Comment (mds1) [inline src/StdUtils.sol:66]: Instead of using `abi.encodePacked`, can just use the `log(string,string)` version. Also note the removal of the `:` for consistency since it wasn't there before
```suggestion
        console2_log("Bound result", vm.toString(result));
```

Comment (mds1) [reply src/StdUtils.sol:66]: @ZeroEkkusu feel free to merge after this change

**Comment by ZeroEkkusu** [comment]
Thanks, @JhChoy - good work.

**Pull request merged**
